WcaOnRails/app/models/competition.rb

```diff
@@ -466,22 +466,76 @@ def country
   end
 
   def self.search(query, params: {})
-    sql_query = "%#{query}%"
-    Competition.where("id LIKE :sql_query OR name LIKE :sql_query OR cellName LIKE :sql_query OR cityName LIKE :sql_query OR countryId LIKE :sql_query", sql_query: sql_query).order(year: :desc, month: :desc, day: :desc)
+    competitions = Competition.where(showAtAll: true)
+
+    if params[:country_iso2].present?
+      country = Country.find_by_iso2(params[:country_iso2])
+      if !country
+        raise WcaExceptions::BadApiParameter, "Invalid country_iso2: #{params[:country_iso2]}"
+      end
+      competitions = competitions.where(countryId: country.id)
+    end
+
+    if params[:start].present?
+      start_date = Date.safe_parse(params[:start])
+      if !start_date
+        raise WcaExceptions::BadApiParameter, "Invalid start: #{params[:start]}"
+      end
+      competitions = competitions.where("CAST(CONCAT(year,'-',month,'-',day) as Datetime) >= ?", start_date)
+    end
+
+    if params[:end].present?
+      end_date = Date.safe_parse(params[:end])
+      if !end_date
+        raise WcaExceptions::BadApiParameter, "Invalid end: #{params[:end]}"
+      end
+      competitions = competitions.where("CAST(CONCAT(year,'-',endMonth,'-',endDay) as Datetime) <= ?", end_date)
+    end
+
+    if query.present?
+      sql_query = "%#{query}%"
+      competitions = competitions.where("id LIKE :sql_query OR name LIKE :sql_query OR cellName LIKE :sql_query OR cityName LIKE :sql_query OR countryId LIKE :sql_query", sql_query: sql_query).order(year: :desc, month: :desc, day: :desc)
+    end
+
+    if params[:sort].present?
+      params[:sort].split(",").each do |field|
+        order = nil
+        if field[0] == "-"
+          order = :desc
+          field = field[1..-1]
+        else
+          order = :asc
+        end
```

Comment:
Yeah, that bothered me slightly too. Though about a number of different ways to do it, but I don't think there is a flawless one.

Ultimately, doing this in stateless controller functions is a lost game. The "real" solution, if the API has a lot of changes, is to make separate class(es) for it, where each field is an object. When that day comes, summon me, and I'll show how it's done.